storage/cloudspanner/map_storage.go

```diff
@@ -0,0 +1,373 @@
+		[]interface{}{tx.treeID, index, writeRev, value.LeafHash, leafValue, value.ExtraData})
+
+	stx.BufferWrite([]*spanner.Mutation{m})
+	return nil
+}
+
+// getMapLeaf fetches and returns the MapLeaf stored at the specified index and
+// revision.
+func (tx *mapTX) getMapLeaf(ctx context.Context, revision int64, index []byte) (*trillian.MapLeaf, error) {
+	cols := []string{colLeafIndex, colMapRevision, colLeafHash, colLeafValue, colExtraData}
+	rowKey := spanner.Key{tx.treeID, index}.AsPrefix()
+	var l *trillian.MapLeaf
+	rows := tx.stx.Read(ctx, mapLeafDataTbl, spanner.KeySets(rowKey), cols)
+	err := rows.Do(func(r *spanner.Row) error {
+		// TODO(alcutter): add MapRevision to trillian.MapLeaf
+		var rev int64
+		var leaf trillian.MapLeaf
+		if err := r.Columns(&leaf.Index, &rev, &leaf.LeafHash, &leaf.LeafValue, &leaf.ExtraData); err != nil {
+			return err
+		}
+		// Leaves are stored by descending revision, so the first one we find which
+		// satisfies this condition is good:
+		if rev <= revision {
+			l = &leaf
+			return errFinished
+		}
+		return nil
+	})
+	if err != nil && err != errFinished {
+		glog.Errorf("failed to read MapLeafData row for rev %d index %x: %v", revision, index, err)
+		return nil, err
+	}
+	return l, nil
+}
+
+// Get returns the values associated with indexes, at revision.
+// Returns a slice of MapLeaf structs containing the requested values.
+// The entries in the returned slice are not guaranteed to be in the same order
+// as the corresponding values in indexes.
+// An error will be returned if there is a problem with the underlying
+// storage.
+func (tx *mapTX) Get(ctx context.Context, revision int64, indexes [][]byte) ([]trillian.MapLeaf, error) {
+	// c will carry any retrieved MapLeaves.
+	c := make(chan *trillian.MapLeaf, len(indexes))
+	// errc will carry any errors while reading from spanner, although we'll only
+	// return to the caller the first, if any.
+	errc := make(chan error, len(indexes))
+
+	for _, idx := range indexes {
+		idx := idx
+		go func() {
+			l, err := tx.getMapLeaf(ctx, revision, idx)
+			if err != nil {
+				glog.Errorf("failed to getMapLeafData(rev=%d, index=%x): %v", revision, idx, err)
+				errc <- err
+				return
+			}
+			c <- l
+		}()
+	}
+
+	// Now wait for the goroutines to do their thing.
+	ret := make([]trillian.MapLeaf, 0, len(indexes))
+	for range indexes {
+		select {
+		case err := <-errc:
+			return nil, err
+		case l := <-c:
+			if l != nil {
+				ret = append(ret, *l)
+			}
+		}
+	}
+	return ret, nil
+}
+
+// GetSignedMapRoot returns the SignedMapRoot for revision.
+// An error will be returned if there is a problem with the underlying storage.
+func (tx *mapTX) GetSignedMapRoot(ctx context.Context, revision int64) (trillian.SignedMapRoot, error) {
+	query := spanner.NewStatement(
+		"SELECT t.TreeID, t.TimestampNanos, t.TreeSize, t.RootHash, t.RootSignature, t.TreeRevision, t.TreeMetadata FROM TreeHeads t" +
```

Comment:
Raw / multiline strings are helpful for queries

Reply:
good call, thanks!